src/Formats/SchemaInferenceUtils.cpp

```diff
@@ -1482,6 +1562,22 @@ DataTypePtr makeNullableRecursively(DataTypePtr type)
         return nested_type ? std::make_shared<DataTypeArray>(nested_type) : nullptr;
     }
 
+    if (which.isVariant())
+    {
+        const auto * variant_type = assert_cast<const DataTypeVariant *>(type.get());
+        DataTypes nested_types;
+        for (const auto & nested_type: variant_type->getVariants())
+        {
+            auto is_low_cardinality = nested_type->lowCardinality();
+            auto has_sub_types = nested_type->haveSubtypes();
+            if (!is_low_cardinality && has_sub_types)
```

Comment:
```suggestion
            if (!nested_type->lowCardinality() && nested_type->haveSubtypes())
```
Condition is quite small, no need to create extra bool variables for each part

Reply:
@Avogar I split them up here because the `&&` operator seems to have a [short circuit behaviour](https://stackoverflow.com/questions/5211961/how-does-c-handle-short-circuit-evaluation) in C++. For `if (!nested_type->lowCardinality() && nested_type->haveSubtypes())`  if `lowCardinality()` evaluates false then `haveSubtypes()` won't get checked. 

But probably there is a less ugly looking way to get around this problem.

Reply:
> if lowCardinality() evaluates false then haveSubtypes() won't get checked.

It's actually the other way, when `lowCardinality()` evaluates true, `!lowCardinality()` will be false and `haveSubtypes()` won't be called. But what is the problem with it? It's totally ok and expected.

Reply:
Yes, you are right